### PR TITLE
handle triple slashes in url schema 'file' correctly

### LINKED ISSUE #3173: 'file' urls are improperly normalized
Compiler improperly normalizes filenames that are 'file' uri. In particular is 'host' part is omitted compiler removes mandatory slash that should be present after '://' so uri 
```
file:///a/b/c.txt
```
becomes
```
file://a/b/c
```

This issue is a blocks passing tests in the fork of es6 module loader where TypeScript is added as a fully supported transpiler

### PR DESCRIPTION
**Pull request by vladima**
Fixes #3173

### CODE REVIEW

File: src/compiler/core.ts [hunk 1]
```diff
@@ -459,6 +459,14 @@ module ts {
             if (path.charCodeAt(2) === CharacterCodes.slash) return 3;
             return 2;
         }
+        // Per RFC 1738'file' URI schema has a shape file://<host>/<path>
```
Comment (DanielRosenwasser) [inline src/compiler/core.ts]: Space before `'file'`.

File: src/compiler/core.ts [hunk 2]
```diff
@@ -459,6 +459,14 @@ module ts {
             if (path.charCodeAt(2) === CharacterCodes.slash) return 3;
             return 2;
         }
+        // Per RFC 1738'file' URI schema has a shape file://<host>/<path>
+        // if <host> is omitted then it is assumed that host value is'localhost',
```
Comment (DanielRosenwasser) [inline src/compiler/core.ts]: space before `'localhost'`

File: src/compiler/core.ts [hunk 3]
```diff
@@ -459,6 +459,14 @@ module ts {
             if (path.charCodeAt(2) === CharacterCodes.slash) return 3;
             return 2;
         }
+        // Per RFC 1738'file' URI schema has a shape file://<host>/<path>
+        // if <host> is omitted then it is assumed that host value is'localhost',
+        // however slash after the omitted <host> is not removed.
+        // file:///folder1/file1 - this is correct URI
+        // file://folder2/file2 - this is incorrect URI
```
Comment (DanielRosenwasser) [inline src/compiler/core.ts]: *a* correct URI/*an* incorrect URI

File: src/compiler/core.ts [hunk 4]
```diff
@@ -459,6 +459,14 @@ module ts {
...
+        // Per RFC 1738'file' URI schema has a shape file://<host>/<path>
+        // if <host> is omitted then it is assumed that host value is'localhost',
+        // however slash after the omitted <host> is not removed.
+        // file:///folder1/file1 - this is correct URI
+        // file://folder2/file2 - this is incorrect URI
+        if (path.lastIndexOf("file:///", 0) === 0) {
+            return "file:///".length;
+        }
         let idx = path.indexOf('://');
         if (idx !== -1) return idx + 3
```
Comment (DanielRosenwasser) [inline src/compiler/core.ts]: Can you put this in braces?

Comment (DanielRosenwasser) [inline src/compiler/core.ts]: Also make this `return "://".length;` to be consistent with the above.

File: src/compiler/core.ts [hunk 5]
```diff
@@ -459,6 +459,14 @@ module ts {
             if (path.charCodeAt(2) === CharacterCodes.slash) return 3;
             return 2;
         }
+        // Per RFC 1738'file' URI schema has a shape file://<host>/<path>
+        // if <host> is omitted then it is assumed that host value is'localhost',
+        // however slash after the omitted <host> is not removed.
+        // file:///folder1/file1 - this is correct URI
+        // file://folder2/file2 - this is incorrect URI
+        if (path.lastIndexOf("file:///", 0) === 0) {
```
Comment (DanielRosenwasser) [inline src/compiler/core.ts]: Why `lastIndexOf`? Does `file://file://foo` need to be special-cased in some way?

Comment (vladima) [inline src/compiler/core.ts]: `lastIndexOf` with start position 0 here works as `startsWith` since it searches backwards starting from given position

Comment (DanielRosenwasser) [inline src/compiler/core.ts]: Ah, so it's a faster check.

**Comment by DanielRosenwasser** [comment]
:+1:

**Pull request merged**